### PR TITLE
Migrate Search plugin to cats-effect

### LINKED ISSUE #4253: Migrate search plugin to CE

### PR DESCRIPTION
**Pull request by olivergrabinski**
Closes #4253

### CODE REVIEW

File: delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/ce/CatsScopeInitialization.scala [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import ch.epfl.bluebrain.nexus.delta.sdk.projects.model.Project
+import monix.bio.{IO => BIO}
+
+/**
+  * Lifecycle hook for organization and project initialization. It's meant to be used for plugins to preconfigure an
+  * organization or project, like for example the creation of a default view or setting the appropriate permissions.
+  * Implementations should use a `many[ScopeInitialization]` binding such that all implementation are collected during
+  * the service bootstrapping.
+  */
+trait CatsScopeInitialization {
```
Comment (olivergrabinski) [inline delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/ce/CatsScopeInitialization.scala:18]: I opted for the addition of this trait along with its the helper function as making a change to the ScopeInitialization would have had an impact across many plugins. We could slowly use this one as the migration goes on and then rename it.

File: delta/plugins/search/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/search/SearchScopeInitialization.scala [hunk 2]
```diff
@@ -1,56 +1,54 @@
...
+    toCatsIO(views.create(defaultViewId, project.ref, SearchViewFactory(defaults, config))).void
+      .handleErrorWith {
+        case _: ViewAlreadyExists       => IO.unit
+        case _: ProjectContextRejection => IO.unit
         case rej                        =>
           val str =
-            s"Failed to create the search view for project '${project.ref}' due to '${rej.reason}'."
-          UIO.delay(logger.error(str)) >> IO.raiseError(ScopeInitializationFailed(str))
+            s"Failed to create the search view for project '${project.ref}' due to '${rej.getMessage}'."
+          IO.delay(logger.error(str)) >> IO.raiseError(ScopeInitializationFailed(str))
```
Comment (dantb) [inline delta/plugins/search/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/search/SearchScopeInitialization.scala:45]: Have a sneaking suspicion the cats logger is already effectful, so this delay will mean nothing gets logged?

Comment (dantb) [reply delta/plugins/search/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/search/SearchScopeInitialization.scala:45]: If that is the case really a linter should pick it up!

**Review by dantb — CHANGES_REQUESTED** [review]

**Review by dantb — CHANGES_REQUESTED** [review]
LGTM, only the logger question matters. I am curious whether we can create better walls around the mapping from Monix to CE for this migration to save us time in the future

**Review by dantb — CHANGES_REQUESTED** [review]
LGTM, only the logger question matters. I am curious whether we can create better walls around the mapping from Monix to CE for this migration to save us time in the future

File: delta/plugins/search/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/search/Search.scala [hunk 3]
```diff
@@ -58,7 +60,7 @@ object Search {
       compositeViews
         .list(
           Pagination.OnePage,
-          CompositeViewSearchParams(deprecated = Some(false), filter = v => UIO.pure(v.id == defaultViewId)),
+          CompositeViewSearchParams(deprecated = Some(false), filter = v => IO.pure(v.id == defaultViewId).toUIO),
```
Comment (dantb) [inline delta/plugins/search/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/search/Search.scala:63]: Is this a redundant effect transform? Using the cats `(v.id == defaultViewId)`.pure might work through inference and wouldn't demand any change when that module is also migrated

File: delta/plugins/search/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/search/SearchScopeInitialization.scala (hunk 2 above)
Comment (olivergrabinski) [reply delta/plugins/search/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/search/SearchScopeInitialization.scala:45]: Thanks, not the first time it's missed... there should indeed be a warning

File: delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeProjectionLifeCycle.scala [hunk 4]
```diff
@@ -96,20 +98,20 @@ object CompositeProjectionLifeCycle {
...
-          Task.pure(CompiledProjection.fromTask(view.metadata, TransientSingleNode, task))
+          Task.pure(CompiledProjection.fromTask(view.metadata, TransientSingleNode, task.toUIO))
         }
     }
 
     override def destroyOnIndexingChange(prev: ActiveViewDef, next: CompositeViewDef): Task[Unit] =
       (prev, next) match {
         case (prev, next) if prev.ref != next.ref                                            =>
-          IO.terminate(new IllegalArgumentException(s"Different views were provided: '${prev.ref}' and '${next.ref}'"))
+          BIO.terminate(new IllegalArgumentException(s"Different views were provided: '${prev.ref}' and '${next.ref}'"))
```
[the remark this answers is not in the document]
Comment (olivergrabinski) [reply delta/plugins/composite-views/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/compositeviews/indexing/CompositeProjectionLifeCycle.scala:114]: That was the idea, yep

**Review by imsdu — APPROVED** [review]

File: delta/plugins/search/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/search/Search.scala [hunk 5]
```diff
@@ -58,7 +60,7 @@ object Search {
       compositeViews
         .list(
           Pagination.OnePage,
-          CompositeViewSearchParams(deprecated = Some(false), filter = v => UIO.pure(v.id == defaultViewId)),
+          CompositeViewSearchParams(deprecated = Some(false), filter = v => IO.delay(v.id == defaultViewId).toUIO),
```
Comment (imsdu) [inline delta/plugins/search/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/search/Search.scala:63]: `UIO.pure` for now is ok waiting for the migration of the search params

**Review by imsdu — APPROVED** [review]

**Review by imsdu — APPROVED** [review]

**Review by dantb — APPROVED** [review]
Nice :)

**Pull request merged**